src/app/icd/client/DefaultICDClientStorage.cpp

```diff
@@ -488,10 +493,29 @@ CHIP_ERROR DefaultICDClientStorage::DeleteAllEntries(FabricIndex fabricIndex)
     return mpClientInfoStore->SyncDeleteKeyValue(DefaultStorageKeyAllocator::FabricICDClientInfoCounter(fabricIndex).KeyName());
 }
 
-CHIP_ERROR DefaultICDClientStorage::ProcessCheckInPayload(const ByteSpan & payload, ICDClientInfo & clientInfo)
+CHIP_ERROR DefaultICDClientStorage::ProcessCheckInPayload(const ByteSpan & payload, ICDClientInfo & clientInfo, bool & refreshKey)
 {
-    // TODO: Need to implement default decription code using CheckinMessage::ParseCheckinMessagePayload
-    return CHIP_NO_ERROR;
+    uint32_t counter;
+    MutableByteSpan appData;
+    auto * iterator = IterateICDClientInfo();
+    while (iterator->Next(clientInfo))
+    {
+        if (CHIP_NO_ERROR ==
+            chip::Protocols::SecureChannel::CheckinMessage::ParseCheckinMessagePayload(clientInfo.shared_key, payload, counter,
+                                                                                       appData))
+        {
+            auto checkInCounter = (counter - clientInfo.start_icd_counter) % kCheckInCounterMax;
+            VerifyOrReturnError(checkInCounter > clientInfo.offset, CHIP_ERROR_INVALID_SIGNATURE);
+            clientInfo.offset = counter - clientInfo.start_icd_counter;
+            if (checkInCounter > kCheckInRolloverConstant)
+            {
+                refreshKey = true;
+            }
```

Comment:
This should go in the CheckInMessage file since it implements the Check-In message requirements.
This will make unit testing this much easier also.

Reply:
I'm leaving it here for now until Mathieu moves the logic into CheckInMessage.cpp.